### PR TITLE
Update the nix build configuration.

### PR DESCRIPTION
**Pull request by farcaller**
Remove the old shell.nix with some legacy versions pinned and replace it with a flake-based shell. It installs rust via rustup instead of fenix to be more generally compatible with the guidelines.

This also adds the rust-toolchain.toml spec with all the components required for wasm, and everything else to make rust-analyzer & clippy happy.

I verified that `cargo build --release` works for the node within the direnv shell created from the flake.

✄ -----------------------------------------------------------------------------

Thank you for your Pull Request! 🙏

Before you submit, please check that:

- [x] **Description:** You added a brief description of the PR, e.g.:
  - What does it do?
  - What important points should reviewers know?
  - Is there something left for follow-up PRs?
- [ ] **Labels:** You labeled the PR appropriately if you have permissions to do so:
  - [ ] `A*` for PR status (**one required**)
  - [ ] `B*` for changelog (**one required**)
  - [ ] `C*` for release notes (**exactly one required**)
  - [ ] `D*` for various implications/requirements
  - [ ] Github project assignment
- [ ] **Related Issues:** You mentioned a related issue if this PR is related to it, e.g. `Fixes #228` or `Related #1337`.
- [ ] **2 Reviewers:** You asked at least two reviewers to review. If you aren't sure, start with GH suggestions.
- [x] **Style Guide:** Your PR adheres to [the style guide](https://github.com/paritytech/substrate/blob/master/docs/STYLE_GUIDE.md)
  - In particular, mind the maximal line length of 100 (120 in exceptional circumstances).
  - There is no commented code checked in unless necessary.
  - Any panickers in the runtime have a proof or were removed.
- [x] **Runtime Version:** You bumped the runtime version if there are breaking changes in the **runtime**.
- [x] **Docs:** You updated any rustdocs which may need to change.
- [ ] **Polkadot Companion:** Has the PR altered the external API or interfaces used by Polkadot?
  - [ ] If so, do you have the corresponding Polkadot PR ready?
  - [ ] Optionally: Do you have a corresponding Cumulus PR?

Refer to [the contributing guide](https://github.com/paritytech/substrate/blob/master/docs/CONTRIBUTING.adoc) for details.

After you've read this notice feel free to remove it.
Thank you!

✄ -----------------------------------------------------------------------------

### CODE REVIEW

File: bin/node-template/flake.nix [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    let
+      pkgs = import nixpkgs { inherit system; };
+    in
+    {
+      devShells.default = pkgs.mkShell {
+        packages = with pkgs; [
+          rustup
+          clang
+          protobuf
+          gnumake
```
Comment (bkchr) [inline bin/node-template/flake.nix:17]: We don't need make?

**Review by bkchr — APPROVED** [review]

File: bin/node-template/flake.nix (hunk 1 above)
Comment (farcaller) [reply bin/node-template/flake.nix]: Here's a funny one. It's already in the `mkShell`. I used devenv.sh initially and frontier actually needs make to be built. When I was preparing the pull request, I cleaned up extra deps and removed devenv as well (as it seemed a simple mkShell would do the trick just fine). But mkShell pulls the basic builders implicitly, so make is still part of the derivation.

TL;DR: we don't need make here. We still need make to build frontier, but this derivation is universal and will contain make.

File: bin/node-template/flake.nix [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
+    flake-utils.url = "github:numtide/flake-utils";
+  };
+
+  outputs = { self, nixpkgs, flake-utils, ... }: flake-utils.lib.eachDefaultSystem (system:
+    let
+      pkgs = import nixpkgs { inherit system; };
+    in
+    {
+      devShells.default = pkgs.mkShell {
```
[the remark this answers is not in the document]
Comment (farcaller) [reply bin/node-template/flake.nix:12]: Fixed, thanks!

File: bin/node-template/flake.nix [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        packages = with pkgs; [
+          rustup
+          clang
+          protobuf
+          rustfmt
+        ];
+        shellHook = ''
+          export LIBCLANG_PATH="${pkgs.libclang.lib}/lib"
```
[the remark this answers is not in the document]
Comment (farcaller) [reply bin/node-template/flake.nix]: Fixed, thanks!

**Review by andresilva — APPROVED** [review]

**Comment by andresilva** [comment]
The `node-template` README mentions using `lorri`, it should be updated to point to `nix-direnv` which this PR uses.

**Comment by bkchr** [comment]
> The `node-template` README mentions using `lorri`, it should be updated to point to `nix-direnv` which this PR uses.

@farcaller please do this and then we can merge this!

**Comment by farcaller** [comment]
@bkchr done. I also added the top-level configuration for substrate as a whole for those who will hack on the project itself (like me) and want the rust-analyzer to work out of the box in nix.

**Comment by bkchr** [comment]
Yeah just tested it, could you maybe for now disable it? Or keep the nix file and rename the `.envrc` to `.envrc_template`

**Comment by farcaller** [comment]
Will do. I’ll add .envrc to the ignores too, then.

**Comment by farcaller** [comment]
Updated per comments.

File: rust-toolchain.toml [hunk 4]
```diff
@@ -0,0 +1,14 @@
+[toolchain]
+channel = "nightly"
```
Comment (ggwpez) [inline rust-toolchain.toml:2]: Uhm… why do we have a root toolchain file now (https://github.com/paritytech/substrate/pull/13580#issuecomment-1465012700)?  
And why is it set to nightly?

Comment (farcaller) [reply rust-toolchain.toml:2]: Hm. It's set to nightly because the docs said to use nightly when I authored it. I guess it's easier if I split it and restore the original scope of the PR that only modifies the template?

Comment (bkchr) [reply rust-toolchain.toml:2]: > Hm. It's set to nightly because the docs said to use nightly when I authored it. I guess it's easier if I split it and restore the original scope of the PR that only modifies the template?

Yes please let's do this for now.

File: bin/node-template/flake.nix [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    let
+      pkgs = import nixpkgs { inherit system; };
+    in
+    {
+      devShells.default = pkgs.mkShell {
+        packages = with pkgs; [
+          rustup
+          clang
+          protobuf
+          rustfmt
```
Comment (bkchr) [inline bin/node-template/flake.nix:17]: You install here `rustfmt` and then also via the `rust-toolchain.toml`. One should be removed.

Comment (farcaller) [reply bin/node-template/flake.nix:17]: rustup seems to ignore `rust-toolchain.toml's` furstfmt at times. Seems to be working now so I cleaned it up.

File: rust-toolchain.toml (hunk 4 above)
Comment (farcaller) [reply rust-toolchain.toml]: done.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2823048

**Comment by bkchr** [comment]
bot rebase

**Comment by bkchr** [comment]
bot merge
